### PR TITLE
Enable ability to use custom `JsonSchemaObject` in `JsonSchemaParser`

### PR DESCRIPTION
**Pull request by WilliamJamieson**
As described in #1781, I have a need to extend the `JsonSchemaParser` to handle an "extension" to JSON schema. In that discussion, I noted that I figured out method for achieving this, but it required duplication with slight alteration of code already present in the `JsonSchemaParser`.

This PR is a possible solution to avoiding that issue. It does this by factoring the direct call to the `JsonSchemaObject` out into a class variable which can be overridden to a subclass of the `JsonSchemaObject` by a sub classing the `JsonSchemaParser`. If there are better alternate solutions to this issue feel free to close this PR and suggest them in #1781.

### CODE REVIEW

**Comment by koxudaxi** [comment]
@WilliamJamieson 
Thank you for creating the PR.
Should we add the change to OpenAPIParser too?

I think that is a good approach. Can you please add a brief unittest?

**Comment by WilliamJamieson** [comment]
> Should we add the change to OpenAPIParser too?
Honestly, I have no clue if this would be useful for OpenAPI or not. Someone who has experience with it might be a better judge of that.

> Can you please add a brief unittest?
I will see what I can do. Please let me know if you have any thoughts for minimally testing this.

**Comment by WilliamJamieson** [comment]
I made an extremely contrived example for this.

**Comment by WilliamJamieson** [comment]
Darn it looks like it doesn't work for Pydantic v1. I'll have to think about this.

**Comment by koxudaxi** [comment]
@WilliamJamieson 
pydantic v1's union behavior is not same as v2.  https://docs.pydantic.dev/1.10/usage/types/#unions

File: datamodel_code_generator/parser/jsonschema.py [hunk 1]
```diff
@@ -367,6 +367,7 @@ def _get_type(type_: str, format__: Optional[str] = None) -> Types:
 @snooper_to_methods(max_variable_length=None)
 class JsonSchemaParser(Parser):
     SCHEMA_PATHS: ClassVar[List[str]] = ['#/definitions', '#/$defs']
+    SCHEMA_OBJECT_TYPE: ClassVar[Type[JsonSchemaObject]] = JsonSchemaObject
```
Comment (koxudaxi) [inline datamodel_code_generator/parser/jsonschema.py:370]: @WilliamJamieson 
Why not just pass it through init and make it an instance variable?
This is not a criticism but a simple question.

Comment (WilliamJamieson) [reply datamodel_code_generator/parser/jsonschema.py:370]: There is no particular reason. My application already required me to make a subclass of the parser itself to introduce some other behaviors that I need. So it makes no difference to me whether it's passed in through the `__init__` or a class variable.

I figured this type of change could drastically change the parser behavior. So forcing users to do it via a subclass made sense. L

**Comment by WilliamJamieson** [comment]
> Darn it looks like it doesn't work for Pydantic v1. I'll have to think about this. Honestly, I have no clue why it doesn't work there as I have only ever used Pydantic v2.

I think the issue is my use of `model_post_init`. That does not appear to be documented in Pydantic v1. Maybe switching it to using `__post_init__` might work? I'll try to investigate this in the morning.

File: tests/parser/test_jsonschema.py [hunk 2]
```diff
@@ -464,3 +467,72 @@ def test_no_additional_imports():
...
+    """
+    Contrived example to extend the JsonSchemaParser to support an alt_type, which
+    replaces the type if present.
+    """
+
+    class AltJsonSchemaObject(JsonSchemaObject):
+        properties: Optional[Dict[str, Union[AltJsonSchemaObject, bool]]] = None
+        alt_type: Optional[str] = None
+
+        def model_post_init(self, context: Any) -> None:
```
Comment (WilliamJamieson) [inline tests/parser/test_jsonschema.py:526]: > I think the issue is my use of `model_post_init`. That does not appear to be documented in Pydantic v1. Maybe switching it to using `__post_init__` might work? I'll try to investigate this in the morning.

This was not entirely correct, I can fix the failure in Pydantic 1.7+, via 
```suggestion
        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)
```
However, I run into pydantic/pydantic#1985 on versions older than that. Adjusting the skip down to `< 1.7` does not prevent the failure in this case as it occurs during test collection not test running.

So its probably best to leave the test as is, and keep the skip.

Comment (koxudaxi) [reply tests/parser/test_jsonschema.py:526]: @WilliamJamieson 
Thank you for your research.
Perhaps this is not a problem with the code generator, but with how Pydantic is used.
It should be the responsibility of the creator of the custom schema model.

**Review by koxudaxi — APPROVED** [review]
LGTM

File: datamodel_code_generator/parser/jsonschema.py (hunk 1 above)
Comment (koxudaxi) [reply datamodel_code_generator/parser/jsonschema.py:370]: make sense!!

**Comment by koxudaxi** [comment]
Thank you very much!!

**Comment by WilliamJamieson** [comment]
@koxudaxi thanks so much for accepting my pull request.

**Pull request merged**